File: architecture-skills/object-oriented-programming-paradigm/examples/interfaces-and-abstract-classes/interfaces-and-abstract-classes.php
<?php
/**
 * Interfaces & abstract classes in PHP.
 *
 * An interface is a pure contract; an abstract class mixes implemented
 * and abstract members. TabularExporter captures shared line-assembly
 * logic once; JsonExporter satisfies the contract directly.
 */

declare(strict_types=1);

final readonly class Metric
{
    public function __construct(
        public string $name,
        public float $value,
    ) {}
}

interface Exporter
{
    public function contentType(): string;
    public function exportAll(array $metrics): string;

    public function fileName(): string
    {
        $ct = $this->contentType();
        return 'report.' . substr($ct, strrpos($ct, '/') + 1);
    }
}

abstract class TabularExporter implements Exporter
{
    public function exportAll(array $metrics): string
    {
        $lines = [$this->header()];
        foreach ($metrics as $metric) {
            $lines[] = $this->row($metric);
        }
        $lines[] = '# ' . count($metrics) . ' records';
        return implode("\n", $lines);
    }

    abstract protected function header(): string;
    abstract protected function row(Metric $metric): string;
}

final class CsvExporter extends TabularExporter
{
    public function contentType(): string { return 'text/csv'; }
    protected function header(): string { return 'name,value'; }
    protected function row(Metric $m): string
    {
        return sprintf('%s,%.2f', $m->name, $m->value);
    }
}

final class JsonExporter implements Exporter
{
    public function contentType(): string { return 'application/json'; }

    public function exportAll(array $metrics): string
    {
        $entries = array_map(
            fn(Metric $m) => sprintf('{"name": "%s", "value": %.2f}', $m->name, $m->value),
            $metrics
        );
        return '[' . implode(', ', $entries) . ']';
    }
}

function deliver(Exporter $exporter, array $metrics): void
{
    echo "--- {$exporter->contentType()} -> {$exporter->fileName()} ---\n";
    echo $exporter->exportAll($metrics), "\n\n";
}

$metrics = [
    new Metric('cpu', 0.93),
    new Metric('memory', 0.67),
    new Metric('disk', 0.41),
];

deliver(new CsvExporter(), $metrics);
deliver(new JsonExporter(), $metrics);

File: architecture-skills/object-oriented-programming-paradigm/examples/interfaces-and-abstract-classes/interfaces-and-abstract-classes.cs
// Interfaces & abstract classes in C#.
//
// An interface is a pure contract; an abstract class mixes implemented
// and abstract members. TabularExporter captures shared line-assembly
// logic once; JsonExporter satisfies the contract directly.

using System.Text;

record Metric(string Name, double Value);

interface IExporter
{
    string ContentType { get; }
    string ExportAll(IReadOnlyList<Metric> metrics);

    string FileName() => "report." + ContentType.Split('/').Last();
}

abstract class TabularExporter : IExporter
{
    public abstract string ContentType { get; }

    public string ExportAll(IReadOnlyList<Metric> metrics)
    {
        var lines = new List<string> { Header() };
        lines.AddRange(metrics.Select(Row));
        lines.Add($"# {metrics.Count} records");
        return string.Join('\n', lines);
    }

    protected abstract string Header();
    protected abstract string Row(Metric metric);
}

sealed class CsvExporter : TabularExporter
{
    public override string ContentType => "text/csv";
    protected override string Header() => "name,value";
    protected override string Row(Metric m) => $"{m.Name},{m.Value:F2}";
}

sealed class JsonExporter : IExporter
{
    public string ContentType => "application/json";

    public string ExportAll(IReadOnlyList<Metric> metrics)
    {
        var entries = string.Join(", ",
            metrics.Select(m => $"{{\"name\": \"{m.Name}\", \"value\": {m.Value:F2}}}"));
        return $"[{entries}]";
    }
}

static void Deliver(IExporter exporter, IReadOnlyList<Metric> metrics)
{
    Console.WriteLine($"--- {exporter.ContentType} -> {exporter.FileName()} ---");
    Console.WriteLine(exporter.ExportAll(metrics));
    Console.WriteLine();
}

var metrics = new List<Metric>
{
    new("cpu", 0.93),
    new("memory", 0.67),
    new("disk", 0.41),
};

Deliver(new CsvExporter(), metrics);
Deliver(new JsonExporter(), metrics);
